### DISCUSSION TITLE
subscribeWithSelector gives union types  instead of individual types

### DISCUSSION THREAD

**Discussion by jff1625** [pmndrs/zustand#2713] [Bug report] [answered]
### Bug Description

![image](https://github.com/user-attachments/assets/9a9ff482-7f96-431e-8db3-9f409a0bdb80)

I'm expecting the correct types to be derived in the selector at line 22, so if I hover `myString` I should see `(parameter) myString: string`

What I see instead is `(parameter) myString: string | boolean`
![image](https://github.com/user-attachments/assets/45e506ca-3603-4ecc-86fc-909855ebcbc2)

Zustand is taking the types of all the values in the array, making a union type of them, and applying that to every value. I would expect instead that each value would have its correct type.

Of course the types are still correct at runtime as shown by the console.log's, but at compile-time TS can get rather angry about it.

### Reproduction Link

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgZwK4CNkGMrHQUwHVgYALAZXwBt8sZo4BfOAMyghDgCIAvVZGAEMAdgBMA9CGCjRNAO6Co+LgG4AsAChQkWIjg58gmPnL0lTVu069+QseIBuI4FSqDVmzcGHGoLQVj4cACyAJ6m0EEImnBwIOEwuMIA5gBcKIneyeoasfEAQhAQVOnoRTQiOYyeGlgQwgJw3iTAglSmRvjpYRHmALyIMXEJSWncXAA0QwXl6Ymo+FMajDmadQ3w8b1BAwad2wAUQ2iYOHhEJBTUtGYHBwCUcH0AfHAH0bmxcAB0v80wrXaQmMS1ijHu90091WGi2Znw3xO2FwBCOnwOAk6jxecAA2pjjN84aMJhlOkTQoVigBdUFvXHErKkmY07GvD5ffT1ZDFBFUCDJA5cRkpSZwGChMD4CAsYamElyzIpaFDWLrHk0b78wXCynlMUSqUy4ZUqjMvXFFWfaoaK1rbnwJDIfAwDrGCwDOGRHLO13A-DvRWjdJcUjUflilklVhtZ1MK1AA

**Comment by dai-shi** [comment, marked as answer]
`as const` should help.
<img width="539" alt="image" src="https://github.com/user-attachments/assets/82d12540-4393-476a-8dc6-26a8305356f5">
https://tsplay.dev/WkeQLN
